File: 프로그래머스 고득점 kit/깊이-너비 우선 탐색/타겟 넘버.py
# bfs 백트래킹 조건 없으니 전부 완탐
from collections import deque, Counter


def solution(numbers, target):
    q = deque([(0, 0)])
    end = []

    while q:
        acc, index = q.popleft()

        if index == len(numbers):
            end.append(acc)
            continue

        q.append((acc + numbers[index], index + 1))
        q.append((acc - numbers[index], index + 1))

    return Counter(end)[target]


print(solution([1, 1, 1, 1, 1],	3) == 5)
print(solution([4, 1, 2, 1], 4) == 2)

File: 프로그래머스 고득점 kit/깊이-너비 우선 탐색/타겟 넘버.js
function solution(numbers, target) {
  let ans = 0

  /** @param {number[]} currNumbers */
  function dfs(currNumbers, acc) {
    if (currNumbers.length === 0) {
      if (acc === target) ans += 1
      return
    }

    const [curr, ...nextNumbers] = currNumbers
    dfs(nextNumbers, acc + curr)
    dfs(nextNumbers, acc - curr)
  }
  dfs(numbers, 0)

  return ans
}

console.log(solution([1, 1, 1, 1, 1], 3) === 5)
console.log(solution([4, 1, 2, 1], 4) === 2)
